File: src/main/java/com/chattriggers/ctjs/utils/console/Console.java
package com.chattriggers.ctjs.utils.console;

import com.chattriggers.ctjs.CTJS;
import com.chattriggers.ctjs.minecraft.libs.RenderLib;
import com.chattriggers.ctjs.triggers.OnTrigger;
import io.sentry.Sentry;
import lombok.Getter;
import net.minecraft.network.ThreadQuickExitException;

import javax.script.ScriptException;
import javax.swing.*;
import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.event.WindowEvent;
import java.io.PrintStream;
import java.util.ArrayList;

public class Console {
    private JFrame frame;
    public PrintStream out;
    private TextAreaOutputStream taos;
    private ArrayList<Component> components;

    @Getter
    private static Console console;

    public Console() {
        this.frame = new JFrame("ct.js Console");
        this.frame.setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE);

        JTextArea textArea = new JTextArea();
        this.taos = new TextAreaOutputStream(textArea, 1000);
        textArea.setEditable(false);
        textArea.setFont(new Font("DejaVu Sans Mono", Font.PLAIN, 15));
        JTextField inputField = new JTextField(1);
        inputField.setFocusable(true);

        inputField.setMargin(new Insets(5, 5, 5, 5));
        textArea.setMargin(new Insets(5, 5, 5, 5));

        components = new ArrayList<>();
        components.add(textArea);
        components.add(inputField);

        inputField.addKeyListener(new KeyListener() {
            @Override
            public void keyTyped(KeyEvent e) {
            }

            @Override
            public void keyPressed(KeyEvent e) {
            }

            @Override
            public void keyReleased(KeyEvent e) {
                if (e.getKeyCode() == KeyEvent.VK_ENTER) {
                    Object toPrint = null;
                    boolean errorThrown = false;

                    String command = inputField.getText();
                    inputField.setText("");

                    try {
                        toPrint = CTJS.getInstance().getModuleManager().eval(command);
                    } catch (Exception error) {
                        if (!(error instanceof ThreadQuickExitException)) {
                            printStackTrace(error);
                            errorThrown = true;
                        }
                    }

                    if (toPrint != null) {
                        out.println(toPrint);
                    } else {
                        out.println(errorThrown ? "> " + command : command);
                    }
                }
            }
        });

        this.out = new PrintStream(taos);

        frame.add(new JScrollPane(textArea));
        frame.add(inputField, BorderLayout.SOUTH);
        frame.pack();
        frame.setVisible(false);
        frame.setSize(800, 600);

        console = this;
    }

    public void clearConsole() {
        this.taos.clearLog();
        this.taos.clear();
    }

    public void printStackTrace(Throwable error) {
        Sentry.capture(error);
        error.printStackTrace(out);
    }

    public void printStackTrace(Throwable error, OnTrigger trigger) {
        Sentry.getContext().addTag(
            "methodName",
            trigger.getMethodName()
        );

        try {
            String body = CTJS.getInstance().getModuleManager().eval(trigger.getMethodName()).toString();

            Sentry.getContext().addExtra(
                "methodBody",
                body.replace("\n", "<br>")
            );
        } catch (ScriptException e) {
            e.printStackTrace();
        }

        if (trigger.getOwningModule() != null) {
            Sentry.getContext().addTag(
                "moduleName",
                trigger.getOwningModule().getName()
            );

            if (trigger.getOwningModule().getMetadata() != null) {
                Sentry.getContext().addExtra(
                    "moduleMetadata",
                    trigger.getOwningModule().getMetadata().toString()
                );
            }
        }

        printStackTrace(error);
    }

    public void showConsole(boolean show) {
        this.frame.setVisible(show);

        Color bg = new Color(29, 31, 33);
        Color fg = new Color(197, 200, 198);

        if (CTJS.getInstance().getConfig().getCustomTheme()) {
            String bgColor = CTJS.getInstance().getConfig().getBg();
            bgColor = bgColor.substring(1, bgColor.length() - 1);
            String[] bgColors = bgColor.split(",");

            bg = new Color(RenderLib.limit255(Integer.parseInt(bgColors[0].trim())),
                    RenderLib.limit255(Integer.parseInt(bgColors[1].trim())),
                    RenderLib.limit255(Integer.parseInt(bgColors[2].trim())));

            String fgColor = CTJS.getInstance().getConfig().getFg();
            fgColor = fgColor.substring(1, fgColor.length() - 1);
            String[] fgColors = fgColor.split(",");

            fg = new Color(RenderLib.limit255(Integer.parseInt(fgColors[0].trim())),
                    RenderLib.limit255(Integer.parseInt(fgColors[1].trim())),
                    RenderLib.limit255(Integer.parseInt(fgColors[2].trim())));
        } else {
            switch (CTJS.getInstance().getConfig().getConsoleTheme()) {
                case "default.dark":
                    bg = new Color(21, 21, 21);
                    fg = new Color(208, 208, 208);
                    break;
                case "ashes.dark":
                    bg = new Color(28, 32, 35);
                    fg = new Color(199, 204, 209);
                    break;
                case "atelierforest.dark":
                    bg = new Color(28, 32, 35);
                    fg = new Color(199, 204, 209);
                    break;
                case "isotope.dark":
                    bg = new Color(0, 0, 0);
                    fg = new Color(208, 208, 208);
                    break;
                case "codeschool.dark":
                    bg = new Color(22, 27, 29);
                    fg = new Color(126, 162, 180);
                    break;
                case "gotham":
                    bg = new Color(10, 15, 20);
                    fg = new Color(152, 209, 206);
                    break;
                case "hybrid":
                    bg = new Color(29, 31, 33);
                    fg = new Color(197, 200, 198);
                    break;
                case "3024.light":
                    bg = new Color(247, 247, 247);
                    fg = new Color(74, 69, 67);
                    break;
                case "chalk.light":
                    bg = new Color(245, 245, 245);
                    fg = new Color(48, 48, 48);
                    break;
                case "blue":
                    bg = new Color(15, 18, 32);
                    fg = new Color(221, 223, 235);
                    break;
                case "slate":
                    bg = new Color(33, 36, 41);
                    fg = new Color(193, 199, 208);
                    break;
                case "red":
                    bg = new Color(26, 9, 11);
                    fg = new Color(231, 210, 212);
                    break;
                case "green":
                    bg = new Color(6, 10, 10);
                    fg = new Color(47, 227, 149);
                    break;
                case "aids":
                    bg = new Color(251, 251, 28);
                    fg = new Color(192, 20, 214);
                    break;
                default:
                    bg = new Color(21, 21, 21);
                    fg = new Color(208, 208, 208);
                    break;
            }
        }

        for (Component comp : this.components) {
            comp.setBackground(bg);
            comp.setForeground(fg);
        }

        this.frame.toFront();
        this.frame.repaint();
    }

    @Override
    public void finalize() {
        this.frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
        this.frame.dispatchEvent(new WindowEvent(frame, WindowEvent.WINDOW_CLOSING));
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/utils/console/Console.kt
package com.chattriggers.ctjs.utils.console

import com.chattriggers.ctjs.Reference
import com.chattriggers.ctjs.engine.ILoader
import com.chattriggers.ctjs.utils.Config
import org.fife.ui.rsyntaxtextarea.RSyntaxTextArea
import org.fife.ui.rsyntaxtextarea.SyntaxConstants
import org.fife.ui.rsyntaxtextarea.Theme
import java.awt.*
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter
import javax.swing.*
import javax.swing.text.DefaultCaret

class Console(val loader: ILoader?) {
    private val frame = JFrame(
        "ChatTriggers ${Reference.MODVERSION} ${loader?.getLanguage()?.langName ?: "Default"} Console"
    )

    private val textArea = JTextPane()
    private val inputField = RSyntaxTextArea(5, 1).apply {
        syntaxEditingStyle = loader?.getLanguage()?.syntaxStyle ?: SyntaxConstants.SYNTAX_STYLE_NONE
        Theme.load(javaClass.getResourceAsStream("/org/fife/ui/rsyntaxtextarea/themes/dark.xml")).apply(this)
        margin = Insets(5, 5, 5, 5)
        isCodeFoldingEnabled = true
    }

    private val components = mutableSetOf<Component>(textArea)
    private val history = mutableListOf<String>()
    private var historyOffset = 0

    val writer = TextAreaWriter(textArea)

    init {
        frame.defaultCloseOperation = JFrame.HIDE_ON_CLOSE

        textArea.isEditable = false

        textArea.margin = Insets(5, 5, 5, 5)
        textArea.autoscrolls = true
        val caret = textArea.caret as DefaultCaret
        caret.updatePolicy = DefaultCaret.ALWAYS_UPDATE

        inputField.addKeyListener(object : KeyListener {
            override fun keyTyped(e: KeyEvent) {}

            override fun keyPressed(e: KeyEvent) {}

            override fun keyReleased(e: KeyEvent) {
                if (!e.isControlDown) return

                when (e.keyCode) {
                    KeyEvent.VK_ENTER -> {
                        val command = inputField.text
                        inputField.text = ""
                        history.add(command)
                        historyOffset = 0

                        if (command == "help") {
                            writer.println("""
                                -------------- ChatTriggers Console Help --------------
                                 Shortcuts:
                                  Control + Enter: Run code in the textbox
                                  Control + UP / DOWN: Cycle between ran code history
                                  Control + L: Clear console
                                  Control + R: Reload ChatTriggers
                                -------------------------------------------------------
                            """.trimIndent())
                        } else {
                            writer.println("eval> ${command.prependIndent("    > ").substring(6)}")

                            try {
                                writer.println(loader?.eval(command) ?: return)
                            } catch (e: Throwable) {
                                printStackTrace(e)
                            }
                        }
                    }

                    KeyEvent.VK_UP -> {
                        historyOffset++

                        try {
                            val message = history[history.size - historyOffset]
                            inputField.text = message
                        } catch (exception: Exception) {
                            historyOffset--
                        }
                    }

                    KeyEvent.VK_DOWN -> {
                        historyOffset--

                        if (historyOffset < 0) historyOffset = 0

                        try {
                            val message = history[history.size - historyOffset]
                            inputField.text = message
                        } catch (exception: Exception) {
                            historyOffset = 0
                            inputField.text = ""
                        }
                    }

                    KeyEvent.VK_L -> {
                        clearConsole()
                    }

                    KeyEvent.VK_R -> {
                        Reference.loadCT()
                    }
                }
            }
        })

        frame.add(JScrollPane(textArea))
        frame.add(inputField, BorderLayout.SOUTH)
        frame.pack()
        frame.isVisible = false
        frame.setSize(800, 600)
    }

    fun clearConsole() {
        SwingUtilities.invokeLater {
            writer.clear()
        }
    }

    @JvmOverloads
    fun println(obj: Any, logType: LogType = LogType.INFO, customColor: Color? = null) {
        SwingUtilities.invokeLater {
            try {
                writer.println(obj.toString(), logType, customColor)
            } catch (exception: Exception) {
                println(obj.toString(), logType, customColor)
            }
        }
    }

    fun printStackTrace(error: Throwable) {
        SwingUtilities.invokeLater {
            try {
                if (Config.openConsoleOnError)
                    showConsole()

                val index = error.stackTrace.indexOfFirst {
                    it?.fileName?.lowercase()?.contains("jsloader") ?: false
                }

                error.stackTrace = error.stackTrace.dropLast(error.stackTrace.size - index - 1).map {
                    val fileNameIndex = it.fileName?.indexOf("ChatTriggers/modules/") ?: return@map it
                    val classNameIndex = it.className.indexOf("ChatTriggers_modules_")

                    if (fileNameIndex != -1) {
                        StackTraceElement(
                            it.className.substring(classNameIndex + 21),
                            it.methodName,
                            it.fileName!!.substring(fileNameIndex + 21),
                            it.lineNumber
                        )
                    } else it
                }.toTypedArray()

                printErrorWithColor(error)
            } catch (ignored: Throwable) {
                error.printStackTrace()
            }
        }
    }

    private fun printErrorWithColor(error: Throwable) {
        val sw = StringWriter()
        error.printStackTrace(PrintWriter(sw))
        val stacktrace = sw.toString()

        writer.println(stacktrace, LogType.ERROR)
    }

    fun showConsole() {
        frame.isVisible = true

        val bg: Color
        val fg: Color

        if (Config.customTheme) {
            bg = Config.consoleBackgroundColor
            fg = Config.consoleForegroundColor
        } else {
            when (Config.consoleTheme) {
                "ashes.dark" -> {
                    bg = Color(28, 32, 35)
                    fg = Color(199, 204, 209)
                }
                "atelierforest.dark" -> {
                    bg = Color(28, 32, 35)
                    fg = Color(199, 204, 209)
                }
                "isotope.dark" -> {
                    bg = Color(0, 0, 0)
                    fg = Color(208, 208, 208)
                }
                "codeschool.dark" -> {
                    bg = Color(22, 27, 29)
                    fg = Color(126, 162, 180)
                }
                "gotham" -> {
                    bg = Color(10, 15, 20)
                    fg = Color(152, 209, 206)
                }
                "hybrid" -> {
                    bg = Color(29, 31, 33)
                    fg = Color(197, 200, 198)
                }
                "3024.light" -> {
                    bg = Color(247, 247, 247)
                    fg = Color(74, 69, 67)
                }
                "chalk.light" -> {
                    bg = Color(245, 245, 245)
                    fg = Color(48, 48, 48)
                }
                "blue" -> {
                    bg = Color(15, 18, 32)
                    fg = Color(221, 223, 235)
                }
                "slate" -> {
                    bg = Color(33, 36, 41)
                    fg = Color(193, 199, 208)
                }
                "red" -> {
                    bg = Color(26, 9, 11)
                    fg = Color(231, 210, 212)
                }
                "green" -> {
                    bg = Color(6, 10, 10)
                    fg = Color(47, 227, 149)
                }
                "aids" -> {
                    bg = Color(251, 251, 28)
                    fg = Color(192, 20, 214)
                }
                "default.dark" -> {
                    bg = Color(41, 49, 52)
                    fg = Color(208, 208, 208)
                }
                else -> {
                    bg = Color(21, 21, 21)
                    fg = Color(208, 208, 208)
                }
            }
        }

        for (comp in components) {
            comp.background = bg
            comp.foreground = fg
        }

        frame.toFront()
        frame.repaint()

        val chosenFont =
            if (Config.consoleFiraCodeFont) FIRA_FONT.deriveFont(Config.consoleFontSize.toFloat()) else Font(
                "DejaVu Sans Mono",
                Font.PLAIN,
                15
            ).deriveFont(Config.consoleFontSize.toFloat())

        textArea.font = chosenFont
        // TODO: Ligatures make everything extremely slow for some reason. Is this fixable?
//        val attrs = FIRA_FONT.attributes.apply { (this as MutableMap<TextAttribute, Any>)[TextAttribute.LIGATURES] = TextAttribute.LIGATURES_ON }
//        inputField.font = FIRA_FONT.deriveFont(attrs)
        inputField.font = chosenFont

        frame.toFront()
        frame.repaint()
    }

    companion object {
        val FIRA_FONT: Font = Font.createFont(
            Font.TRUETYPE_FONT,
            System.getProperty("ct.firaFile", null)?.let { File(it).inputStream() }
                ?: this::class.java.getResourceAsStream("/FiraCode-Regular.otf")
        ).deriveFont(9f)

        init {
            GraphicsEnvironment.getLocalGraphicsEnvironment().registerFont(FIRA_FONT)
        }
    }
}
